File: misc/generate-zipfian-words.pl
#!/usr/bin/perl

use strict;
use warnings;
use bytes;

use List::Util qw( shuffle );
use POSIX qw( floor );

use constant A => 1.1;
use constant LINES => 10**6;
use constant WORDS => 3;

# ---- load array ----

my @words;

open(IN, '/usr/share/dict/words') || die($!);

while( <IN> )
{
  chomp;
  push( @words, $_ );
}

# ---- shuffle ----

@words = shuffle(@words);

# ---- generate lines ----

# zipfian variate, \propto x^{-A}
# http://luc.devroye.org/chapter_ten.pdf, p551

my $b = 2**(A-1);

my ($u, $v, $x, $t);

for( my $i = 0; $i < LINES; $i++ )
{
  my $line = '';

  for( my $j = 0; $j < WORDS; $j++ )
  {
    do
    {
      ($u, $v) = (rand(1), rand(1));
      $x = $u**(-1/(A-1));
      $t = (1+1/$x)**(A-1);

    } until( ($v*$x*($t-1)/($b-1)) <= ($t/$b) && $x < scalar(@words) );

    $line .= $words[floor($x)];
  }

  print "$line\n";
}

File: misc/generate-zipfian-words.py
#!/usr/bin/env python2

from math import floor
from random import random

def generateZipfian(input_fn, a, words):

  b = 2**(a-1)

  # populate input dictionary
  dict = [ ]

  with open(input_fn) as input_file:
    for line in input_file:
      dict.append(line.rstrip())

  # generator
  while True:
    ret = ''

    for i in xrange(words):

      while True:
        u = random()
        v = random()
        x = u**(-1/(a-1))
        t = (1+1/x)**(a-1)

        if (v*x*(t-1)/(b-1)) <= (t/b) and x < len(dict):
          break;

      ret += dict[int(floor(x))]

    yield ret

# ---- Main function ----

if __name__ == '__main__':

  z = generateZipfian('/usr/share/dict/words',
                      a = 1.1, words = 3);

  for i in xrange(10**6):
    print(next(z))

